amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -313,8 +319,24 @@ private void errorOccurred(RecordFlow triggeringFlow, Throwable t) {
         }
     }
 
-    private void clearRecordsDeliveryQueue() {
+    // This method is not thread safe. This needs to be executed after acquiring lock on this.lockObject
+    private void clearRecordsDeliveryQueue(boolean isSubscriptionCompleting) {
```

Comment:
Couple comment on this method: 

Why it's necessary to introduce split-brain logic here? Do we expect queue to be full, by the time subscribeToShard call is made? That invocation seems to be backup plan. We can either pull that out to its own method, or just let the start of a new subscription wait as well. 

isAwaitingFinalAck variable is tightly couple with the queue status, why not update that within the if? i.e. if !queue.isEmpty -> then set the variable, and lock.wait()

Reply:
`Do we expect queue to be full, by the time subscribeToShard call is made? ` - Ideally no, just for precaution. 
`We can either pull that out to its own method` - would be preferred instead of waiting
`isAwaitingFinalAck variable is tightly couple with the queue status, why not update that within the if?` - Updates as part of yatins comment